File: Corte_II/02-Retos/Retos-Recursividad/Reto-III.cpp
//Invertir una cadena de texto recursivamente

#include <iostream>
#include <string>

using namespace std;

// Función que invierte un texto de forma recursiva
string invertir_cadena(string cadena) {
    // Caso base: si el texto tiene 0 o 1 carácter, se devuelve tal cual
    if (cadena.length() <= 1) {
        return cadena;
    } else {
        // Caso recursivo: se toma el último carácter y se concatena con la inversión del resto de la cadena
        return cadena.back() + invertir_cadena(cadena.substr(0, cadena.length() - 1));
        //cadena.back(): Obtiene el último carácter del texto actual.
        //cadena.substr(0, cadena.length() - 1): Obtiene una subcadena que excluye el último carácter.
    }
}

int main() {
    string cadena;
    cout << "Ingrese una cadena de texto: ";
    
    getline(cin, cadena); // Se usa getline para leer toda la línea, incluyendo espacios
    
    // Llamal a la función recursiva
    string cadena_invertida = invertir_cadena(cadena);
    
    // Muestra el resultado final en pantalla
    cout << "La cadena invertida es: " << cadena_invertida << endl;
    
    return 0;
}

File: Corte_II/02-Retos/Retos-Recursividad/Reto-III.py
#Invertir una cadena de texto recursivamente

def invertir_cadena(cadena):
    # Caso base: si la cadena está vacía o tiene un solo carácter, se devuelve tal cual
    if len(cadena) <= 1:
        return cadena
    else:
        # Llamada recursiva: se toma el último carácter y se concatena con la inversión del resto de la cadena
        return cadena[-1] + invertir_cadena(cadena[:-1])
        #cadena[-1] Toma la última letra del texto
        #cadena [:-1] Toma el texto sin la última letra

cadena_de_prueba = "Hola"
invertida = invertir_cadena(cadena_de_prueba)
print(f"La cadena original es: {cadena_de_prueba}")
print(f"La cadena invertida es: {invertida}")